### PR TITLE
light: improve timeout functionality

### PR DESCRIPTION
**Pull request by cmwaters**
This PR makes the following changes:

- Reduces the default timeout and retry attempts from 5 seconds and 10 attempts to 3 seconds and 5 attempts. This reduces the total possible time for a light client request from 200 seconds to just over 50 seconds
- Takes advantage of the RPC error changes to only retry a request when a provider hasn't responded
- Adds a timeout to the light rpc server which is equivalent to the write timeout set in the config

### CODE REVIEW

File: cmd/tendermint/commands/light.go [hunk 1]
```diff
@@ -222,6 +217,11 @@ func runProxy(cmd *cobra.Command, args []string) error {
 		cfg.WriteTimeout = config.RPC.TimeoutBroadcastTxCommit + 1*time.Second
 	}
 
+	rpcClient, err := rpchttp.NewWithTimeout(primaryAddr, "/websocket", cfg.WriteTimeout)
```
Comment (alexanderbez) [inline cmd/tendermint/commands/light.go:220]: This is always some default value, right? What is that value?

File: cmd/tendermint/commands/light.go [hunk 2]
```diff
@@ -222,6 +217,11 @@ func runProxy(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		return fmt.Errorf("http client for %s: %w", primaryAddr, err)
```
Comment (alexanderbez) [inline cmd/tendermint/commands/light.go:222]:
```suggestion
		return fmt.Errorf("failed to create http client for %s: %w", primaryAddr, err)
```

File: light/provider/http/http.go [hunk 3]
```diff
@@ -146,29 +147,26 @@ func (p *http) validatorSet(ctx context.Context, height *int64) (*types.Validato
...
-				// time. If this is the case then we continue, else we should stop straight away
-				time.Sleep(backoffTimeout(uint16(attempt)))
-				continue
+				return nil, provider.ErrBadLightBlock{Reason: e}
 
 			default:
-				// something has happened to the RPC module if we are not receiving errors of type RPCError
-				panic(fmt.Errorf("unexpected error type: %w", err))
+				// If we don't know the error then by default we return a bad light block error and terminate the connection
+				// with the peer
```
Comment (alexanderbez) [inline light/provider/http/http.go:168]:
```suggestion
				// If we don't know the error then by default we return a bad light block error and
				// terminate the connection with the peer.
```

File: light/provider/http/http.go [hunk 4]
```diff
@@ -210,15 +216,10 @@ func (p *http) signedHeader(ctx context.Context, height *int64) (*types.SignedHe
...
-			// TODO: If we can, we should check if the error is purely because the node failed to respond in
-			// time. If this is the case then we continue, else we should stop straight away
-			time.Sleep(backoffTimeout(uint16(attempt)))
-			continue
-
 		default:
-			// something has happened to the RPC module if we are not receiving errors of type RPCError
-			panic(fmt.Errorf("unexpected error type: %w", err))
+			// If we don't know the error then by default we return a bad light block error and terminate the connection
+			// with the peer
```
Comment (alexanderbez) [inline light/provider/http/http.go:221]:
```suggestion
			// If we don't know the error then by default we return a bad light block error and
			// terminate the connection with the peer.
```

**Review by alexanderbez — APPROVED** [review]
🎉

**Comment by alexanderbez** [comment]
Should this be backported?

File: cmd/tendermint/commands/light.go (hunk 1 above)
Comment (cmwaters) [reply cmd/tendermint/commands/light.go:220]: 10 seconds

**Comment by cmwaters** [comment]
Nah most of it builds off prior breaking work. We could backport the changes to the timeout and the addition of the timeout in the light cli but I would go with just leaving this till 0.35

**Pull request merged**